src/legacy/server/saved_objects/service/saved_objects_client.ts

```diff
@@ -0,0 +1,307 @@
+   * doing substring checks on `error.body.error.reason`, just use the helpers to
+   * understand the meaning of the error:
+   *
+   *   ```js
+   *   if (savedObjectsClient.errors.isNotFoundError(error)) {
+   *      // handle 404
+   *   }
+   *
+   *   if (savedObjectsClient.errors.isNotAuthorizedError(error)) {
+   *      // 401 handling should be automatic, but in case you wanted to know
+   *   }
+   *
+   *   // always rethrow the error unless you handle it
+   *   throw error;
+   *   ```
+   *
+   * ### 404s from missing index
+   *
+   * From the perspective of application code and APIs the SavedObjectsClient is
+   * a black box that persists objects. One of the internal details that users have
+   * no control over is that we use an elasticsearch index for persistance and that
+   * index might be missing.
+   *
+   * At the time of writing we are in the process of transitioning away from the
+   * operating assumption that the SavedObjects index is always available. Part of
+   * this transition is handling errors resulting from an index missing. These used
+   * to trigger a 500 error in most cases, and in others cause 404s with different
+   * error messages.
+   *
+   * From my (Spencer) perspective, a 404 from the SavedObjectsApi is a 404; The
+   * object the request/call was targeting could not be found. This is why #14141
+   * takes special care to ensure that 404 errors are generic and don't distinguish
+   * between index missing or document missing.
+   *
+   * ### 503s from missing index
+   *
+   * Unlike all other methods, create requests are supposed to succeed even when
+   * the Kibana index does not exist because it will be automatically created by
+   * elasticsearch. When that is not the case it is because Elasticsearch's
+   * `action.auto_create_index` setting prevents it from being created automatically
+   * so we throw a special 503 with the intention of informing the user that their
+   * Elasticsearch settings need to be updated.
+   *
+   * @type {ErrorHelpers} see ./lib/errors
+   */
+  public static errors = errors;
+  public errors = errors;
+
+  private _repository: SavedObjectsRepository;
+
+  constructor(repository: SavedObjectsRepository) {
+    this._repository = repository;
+  }
+
+  /**
+   * Persists a SavedObject
+   *
+   * @param type
+   * @param attributes
+   * @param options
+   */
+  async create<T extends SavedObjectAttributes = any>(
+    type: string,
+    attributes: T,
+    options?: CreateOptions
+  ) {
+    return this._repository.create(type, attributes, options);
+  }
+
+  /**
+   * Persists multiple documents batched together as a single request
+   *
+   * @param objects
+   * @param options
+   */
+  async bulkCreate<T extends SavedObjectAttributes = any>(
+    objects: Array<BulkCreateObject<T>>,
+    options?: CreateOptions
+  ) {
+    return this._repository.bulkCreate(objects, options);
```

Comment:
optional: we can skip `await`, although I noticed that most developer declare async operations explicitly. should we consider it as a convention?